File: android/src/main/java/com/netrareactnative/dto/OfflinePolicyActionDTO.kt
package com.example.netra_flutter.dto

import com.netra.library.enums.OfflinePolicyAction
import kotlin.time.Duration.Companion.milliseconds

data class OfflinePolicyActionDTO(
    val identifier: String,
    val retries: Double?,
    val retryDuration: Double?,
    val retryUnit: String?,
) {
    companion object {
        fun fromDataModel(offlinePolicyAction: OfflinePolicyAction): OfflinePolicyActionDTO {
            return OfflinePolicyActionDTO(
                identifier = offlinePolicyAction.identifier,
                retries = ((if (offlinePolicyAction is OfflinePolicyAction.RETRY) {
                    offlinePolicyAction.retries
                } else {
                    null
                })?.toDouble()),
                retryDuration = ((if (offlinePolicyAction is OfflinePolicyAction.RETRY) {
                    offlinePolicyAction.retryInterval?.inWholeMilliseconds
                } else {
                    null
                })?.toDouble()),
                retryUnit = ((if (offlinePolicyAction is OfflinePolicyAction.RETRY) {
                    "MILLISECONDS"
                } else {
                    null
                }))
            )
        }
    }

    fun toDataModel(): OfflinePolicyAction {
        return OfflinePolicyAction.fromIdentifier(
            identifier,
            retries?.toInt(),
            retryDuration?.milliseconds,
        )
    }
}

File: src/internal/dto/OfflinePolicyActionDTO.ts
import {
  OfflinePolicyAction,
  RetryPolicyAction,
} from '../../models/OfflinePolicyAction';
import { BaseDTO } from './BaseDTO';

export class OfflinePolicyActionDTO extends BaseDTO {
  identifier: string;
  retries?: number;
  retryDuration?: number;
  retryUnit?: string;

  constructor(
    identifier: string,
    retries?: number,
    retryDuration?: number,
    retryUnit?: string
  ) {
    super();
    this.identifier = identifier;
    this.retries = retries;
    this.retryDuration = retryDuration;
    this.retryUnit = retryUnit;
  }

  static fromDataModel(model: OfflinePolicyAction) {
    return new OfflinePolicyActionDTO(
      model.identifier,
      model instanceof RetryPolicyAction ? model.retries : undefined,
      model instanceof RetryPolicyAction ? model.retryInterval : undefined,
      model instanceof RetryPolicyAction ? 'MILLISECONDS' : undefined
    );
  }

  static fromJSON(json: string) {
    const parsedJSON = JSON.parse(json);
    return new OfflinePolicyActionDTO(
      parsedJSON.identifier,
      parsedJSON.retries,
      parsedJSON.retryInterval,
      parsedJSON.retryUnit
    );
  }

  toDataModel(): OfflinePolicyAction {
    return OfflinePolicyAction.fromIdentifier(
      this.identifier,
      this.retries,
      this.retryDuration
    );
  }

  toJSON() {
    return {
      identifier: this.identifier,
      retries: this.retries,
      retryDuration: this.retryDuration,
      retryUnit: this.retryUnit,
    };
  }
}
